File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.cpp
class RandomizedSet {
private:
    unordered_set<int> stt;
public:
    RandomizedSet() {
        ios_base::sync_with_stdio(false); 
    }
    
    bool insert(int val) {
        if (!stt.count(val)) {
            stt.insert(val);
            return true;
        } else {
            return false;
        }
    }
    
    bool remove(int val) {
        auto it = stt.find(val);
        if (it != stt.end()) {
            stt.erase(it);
            return true;
        } else {
            return false;
        }
    }
    
    int getRandom() {
        int randomIndex = rand() % stt.size();

        // Use an iterator to select the element at the random index
        auto it = next(stt.begin(), randomIndex);

        // Access the randomly chosen element
        int randomElement = *it;

        return randomElement;
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.java
class RandomizedSet {
    private HashMap<Integer, Integer> idxMap;
    private ArrayList<Integer> list;

    public RandomizedSet() {
        idxMap = new HashMap<>();
        list = new ArrayList<>();
    }
    
    public boolean insert(int val) {
        if (idxMap.containsKey(val)) {
            return false;
        }

        list.add(val);
        idxMap.put(val, list.size() - 1);

        return true;
    }
    
    public boolean remove(int val) {
        if (!idxMap.containsKey(val)) {
            return false;
        }

        int removalIdx = idxMap.get(val);
        int lastEl = list.get(list.size() - 1);

        list.set(removalIdx, lastEl);
        list.remove(list.size() - 1);

        idxMap.put(lastEl, removalIdx);

        idxMap.remove(val);

        return true;
    }
    
    public int getRandom() {
        int randomIdx = (int) (Math.random() * list.size());

        return list.get(randomIdx);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */